### PR TITLE
Add Defines Loading

### PR DESCRIPTION
**Pull request by Spartan322**
Caches start_date and end_date in DefineManager

Add static `get_property` and HASID_PROPERTY macro to HasIdentifier

### CODE REVIEW

File: src/openvic-simulation/dataloader/NodeTools.cpp [hunk 1]
```diff
@@ -142,8 +142,8 @@ node_callback_t NodeTools::expect_colour(callback_t<colour_t> callback) {
 	};
 }
 
-node_callback_t NodeTools::expect_date(callback_t<Date> callback) {
-	return expect_identifier(
+node_callback_t NodeTools::expect_date(callback_t<Date> callback, string_conversion_callback_t parse_callback) {
```
Comment (BrickPi) [inline src/openvic-simulation/dataloader/NodeTools.cpp:145]: Solved this problem differently with OOBs, just changed `expect_date` to use `expect_identifier_or_string` which feels more concise to use. This method works too but we need to decide on one to use.

File: src/openvic-simulation/misc/Define.cpp [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+						case 'p': // pops
+							type = Pops;
+							break;
+						case 'a': // ai
+							type = Ai;
+							break;
+						case 'g': // graphics
+							type = Graphics;
+							break;
+						default: return false;
```
Comment (BrickPi) [inline src/openvic-simulation/misc/Define.cpp:100]: Would log an error message with the invalid identifier here.

**Review by BrickPi — CHANGES_REQUESTED** [review]
Looks good generally, `HASID_PROPERTY` is a big clarity improvement & start_date loading is appreciated.

File: src/openvic-simulation/dataloader/NodeTools.cpp (hunk 1 above)
Comment (Spartan322) [reply src/openvic-simulation/dataloader/NodeTools.cpp:145]: Could change the default to be `expect identifier_or_string` and then introduce the parse_callback for any additional behavior instead, that's probably the best solution.

Comment (BrickPi) [reply src/openvic-simulation/dataloader/NodeTools.cpp:145]: Sounds good to me

Comment (Hop311) [reply src/openvic-simulation/dataloader/NodeTools.cpp:145]: The approach I've used for things like this is to return ‘std::string_view‘ callback, so you can then choose to do ‘expect_identifier(expect_date(...))‘ or ‘expect_string(expect_date(...))‘, or any other function which accepts a string_view callback. This also removes the need for a Node as the root source of the date

Comment (Hop311) [reply src/openvic-simulation/dataloader/NodeTools.cpp:145]: Thinking about it some more, every expect function which is basically just # a specialisation of ‘expect_identifier‘ should return a string_view callback, eg expect_fixed_point.

**Review by BrickPi — APPROVED** [review]

**Pull request merged**